### PR TITLE
EOS-16036: Added implementation to copy source object metadata and tags by default to copied object.

### PR DESCRIPTION
**Pull request by amitwac2608**
Added UTs, and fixed one error message.

Signed-off-by: Amit Kumar <amit.kumar@seagate.com>

### CODE REVIEW

File: server/s3_copy_object_action.cc [hunk 1]
```diff
@@ -221,6 +221,43 @@ void S3CopyObjectAction::fetch_source_object_info_failed() {
...
+  auto meta_directive_hdr =
+      request->get_header_value("x-amz-metadata-directive");
+  auto tagging_directive_hdr =
+      request->get_header_value("x-amz-tagging-directive");
+
+  if (!meta_directive_hdr.empty()) {
+    s3_log(S3_LOG_DEBUG, stripped_request_id,
+           "Received x-amz-metadata-directive header, value: %s",
+           meta_directive_hdr.c_str());
+    if (meta_directive_hdr == DirectiveValueReplace) {
```
Comment (evgeniy-brazhnikov) [inline server/s3_copy_object_action.cc:233]: HTTP standard states case insensitivity

File: server/s3_copy_object_action.cc [hunk 2]
```diff
@@ -221,6 +221,43 @@ void S3CopyObjectAction::fetch_source_object_info_failed() {
...
+      s3_put_action_state = S3PutObjectActionState::validationFailed;
+      set_s3_error("NotImplemented");
+      send_response_to_s3_client();
+      return;
+    } else if (meta_directive_hdr != DirectiveValueCOPY) {
```
Comment (evgeniy-brazhnikov) [inline server/s3_copy_object_action.cc:238]: HTTP standard states case insensitivity

File: ut/s3_get_object_action_test.cc [hunk 3]
```diff
@@ -275,6 +275,12 @@ TEST_F(S3GetObjectActionTest, ValidateObjectOfSizeZero) {
       .Times(AtLeast(1));
   EXPECT_CALL(*ptr_mock_request, send_reply_end()).Times(1);
 
+  std::map<std::string, std::string> meta_map{{"key", "value"}};
```
Comment (evgeniy-brazhnikov) [inline ut/s3_get_object_action_test.cc:278]: It possible to see here std::map<>::map(std::initializer_list<const char*>({...})).
Hope, the compiler see what you want.
Suggest to test for ensure

File: server/s3_copy_object_action.cc (hunk 1 above)
Comment (amitwac2608) [reply server/s3_copy_object_action.cc:233]: I tried to use "copy" with AWS endpoint for CopyObject API, it fails with "InvalidArgument":

# aws --**endpoint-url https://s3.amazonaws.com** --profile aws s3api copy-object --bucket amitbk --copy-source archive-drone-pune-west/file-with-tags --key copy3 **--metadata-directive copy**

An error occurred (**InvalidArgument**) when calling the CopyObject operation: Unknown metadata directive.


# aws --endpoint-url https://s3.amazonaws.com --profile aws s3api copy-object --bucket amitbk --copy-source archive-drone-pune-west/file-with-tags --key copy3 **--metadata-directive COPY**
{
    "CopyObjectResult": {
        "LastModified": "2020-12-29T06:45:07.000Z",
        "ETag": "\"0f343b0931126a20f133d67c2b018a3b\""
    }
}

File: server/s3_copy_object_action.cc (hunk 2 above)
Comment (amitwac2608) [reply server/s3_copy_object_action.cc:238]: AWS is mandating the case-sensitivity. I tested it with AWS. Included the response in other comment.

File: ut/s3_get_object_action_test.cc (hunk 3 above)
Comment (amitwac2608) [reply ut/s3_get_object_action_test.cc:278]: yes, UTs have compiled and ran successfully. I have fired Jenkins job as well..

Can you elaborate more on what's the issue here though ?

**Review by bkirunge7 — APPROVED** [review]

**Review by knrajnambiar76 — APPROVED** [review]

**Comment by amitwac2608** [comment]
Jenkins Job passed on latest commit: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/353/

File: server/s3_copy_object_action.cc [hunk 4]
```diff
@@ -526,16 +563,18 @@ void S3CopyObjectAction::save_metadata() {
...
-    }
-  }*/
+
+  // put source object tags on new object
+  s3_log(S3_LOG_DEBUG, stripped_request_id,
+         "copying source object tags on new object..");
+  new_object_metadata->set_tags(source_object_metadata->get_tags());
+
+  // copy source object user-defined metadata, to new object
+  s3_log(S3_LOG_DEBUG, stripped_request_id,
```
[the remark this answers is not in the document]
Comment (amitwac2608) [reply server/s3_copy_object_action.cc:573]: Ok, We can remove this logging in next PR for CopyObject API change. @evgeniy-brazhnikov has to do code changes for a bug and refactoring in this action class.

[the remark this answers is not in the document]
Comment (knrajnambiar76) [reply server/s3_copy_object_action.cc:573]: ok

[the remark this answers is not in the document]
Comment (knrajnambiar76) [reply server/s3_copy_object_action.cc:573]: below also another logging is there that can also be removed

**Pull request merged**